### ISSUE TITLE
filesystem_cache with s3Cluster

### ISSUE THREAD

**Issue by jurajmasar** [ClickHouse/ClickHouse#72816]
[Remote file caching for s3 deployed in 24.10](https://clickhouse.com/blog/clickhouse-release-24-10#caching-remote-files) by @kssenii works great with `s3()` function, but unfortunately it seems to be unusable with multiple hosts in a cluster with `s3Cluster()`.

`s3Cluster()` seems to distribute each file to **a randomly chosen host in the cluster** so if a single data source is queried repeatedly, the cache ends up being fully duplicated on every node.

Could we please adjust `s3Cluster()` so that the distribution of files among hosts is stable, e.i. it keeps asking the same host in the cluster for the same S3 file?

## Sample query to replicate this issue

clickhouse_config.xml
```
<clickhouse>
  <filesystem_caches>
    <s3_cache>
      <path>/var/lib/clickhouse/s3_cache</path>
      <max_size>10Gi</max_size>
    </s3_cache>
  </filesystem_caches>
</clickhouse>
```

```
SELECT *
FROM s3Cluster(primary,'https://datasets-documentation.s3.eu-west-3.amazonaws.com/nyc-taxi/trips_*.gz', 'TabSeparatedWithNames')
FORMAT NULL 
SETTINGS filesystem_cache_name = 's3_cache', enable_filesystem_cache = 1
```

### Current behavior
After repeated runs, **the cache directory on each host** ends up with the entire data set.

### Proposed situation
`s3Cluster()` would keep distributing files among hosts the same way across multiple queries so that the entire dataset would be distributed among all hosts even after multiple query runs.

Thank you! 🙏

**Comment by kssenii** [comment]
Hi @jurajmasar, this issue definitely needs to be fixed, I cannot give currently an estimation about when it will be fixed.

**Comment by alexey-milovidov** [comment]
There are a few possible solutions.

The complex one is - preferring distribution with consistent hashing, but it could be worse for non-uniform file sizes, and non-uniform replicas performance. The tail latency will suffer.

The easy one is - memoizing the previous distribution in memory and preferring to reuse it if the set of hosts didn't change.

**Comment by taytayprime** [comment]
+1

filesystem cache is unfortunately unusable with s3cluster without this

**Comment by tvi** [comment]
+1

Just came across this issue. 

> The complex one is - preferring distribution with consistent hashing, but it could be worse for non-uniform file sizes, and non-uniform replicas performance. The tail latency will suffer.

What about the solution distribution with consistent hashing, but make it opt-in? For use-cases that are not worried about tail latency.

**Comment by alexey-milovidov** [comment]
@tvi, yes, it will be a reasonable approach and actually not hard to do.

**Comment by mscavnicky** [comment]
+1 this is a blocker currently

**Comment by devcrafter** [comment]
Done in https://github.com/ClickHouse/ClickHouse/pull/77326
